### PR TITLE
Handle multi-result operations and dispatches.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
The TileAndDistributeToWorkgroups could only handle cases where fused
producer has a single use. For multiple uses, each use would introduce
new instance of the tiled producer with tile + fuse. To address this
add a `TileAndFuse` pattern that fuses the producer right after
tiling. This allows updating multiple uses of the producer and also
update the `flow.dispatch.tensor.store` operations to use tiled
versions of this operation. This also allows handling dispatches where
multiple results are returned.

Issue #10228

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp [hunk 1]
```diff
@@ -390,45 +475,195 @@ FailureOr<TilingResult> TileDispatchUsingSCFForOp::returningMatchAndRewrite(
...
+    candidateSliceOps =
+        llvm::to_vector(llvm::makeArrayRef(candidateSliceOps).drop_front());
+    return {};
+  }
+  SmallVector<tensor::ExtractSliceOp> slicesFromSameProducer;
+  SmallVector<tensor::ExtractSliceOp> remainingCandidates;
+  slicesFromSameProducer.push_back(sliceOp);
+  for (auto otherSliceOp : llvm::makeArrayRef(candidateSliceOps).drop_front()) {
+    auto currProducer = otherSliceOp.getSource().getDefiningOp();
+    if (currProducer == producer.getOperation()) {
```
Comment (pzread) [inline compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp:559]: Will there be a case where a single producer is used in multiple slice ops, but those slice ops slice the producer's result into different shapes?

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp:559]: In the general case yes. But it is hard to enforce that this doesnt happen and to handle all cases appropriately. The general tile and fuse mechanism will have to assume some properties, like "all producers has compatible parallel loops as the consumer" or not, which then implies that all slices are the same. I can do some checking though. Will add that.

**Comment by hanhanW** [comment]
The benchmark compare the benchmark numbers with ToT, which includes my recent improvements. Would you like to rebase it on main, so we can make sure that there are no regressions?

**Comment by MaheshRavishankar** [comment]
> The benchmark compare the benchmark numbers with ToT, which includes my recent improvements. Would you like to rebase it on main, so we can make sure that there are no regressions?

Oh it isnt rebased on ToT. I thought it was and was trying to looking the regressions. Thanks for checking!

File: compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp:559]: I need to think about what checks I can add here that make sense that work for `TilingInterface`. For now I think we are OK just making this assumption. I'll give this some thought and send a follow up.

File: compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp [hunk 2]
```diff
@@ -390,45 +475,195 @@ FailureOr<TilingResult> TileDispatchUsingSCFForOp::returningMatchAndRewrite(
...
-            resultSizes, resultStrides, combinedOffsets, combinedSizes,
-            combinedStrides))) {
-      return rewriter.notifyMatchFailure(
-          op, "failed to create tiled flow.dispatch.tensor.store op");
+    tensor::ExtractSliceOp sliceOp = currSliceOps.back();
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPoint(sliceOp);
+    FailureOr<Value> tiledProducerVal =
+        tensor::replaceExtractSliceWithTiledProducer(
+            rewriter, currSliceOps.front(), producerVal);
```
Comment (pzread) [inline compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp:614]: nit: Any reason not to use sliceOp instead of currSliceOps.front()?

File: compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp [hunk 3]
```diff
@@ -194,12 +194,82 @@ static SmallVector<scf::ForOp> generateTileLoopNest(
...
+      storeOp.getLoc(), tiledValue, storeOp.getTarget(),
+      storeOp.getTargetDims(), combinedOffsets, combinedSizes, combinedStrides);
+  rewriter.eraseOp(storeOp);
+  return success();
+}
+
+/// Replace all `flow.dispatch.tensor.store` operations that use values produced
+/// by `untiledOp` as source, with tiled stores, with tiled values produced by
+/// `tiledOp`.
+static LogicalResult replaceWithTiledVersionOfStores(
```
Comment (pzread) [inline compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp:240]: nit: Wondering if `replaceAllStoresWithTiledVersion` will be better distinguished from `replaceWithTiledVersionOfStore`

File: compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp [hunk 4]
```diff
@@ -390,45 +475,195 @@ FailureOr<TilingResult> TileDispatchUsingSCFForOp::returningMatchAndRewrite(
...
-    if (failed(IREE::Flow::foldOffsetsSizesAndStrides(
-            rewriter, loc, storeOp.getMixedOffsets(), storeOp.getMixedSizes(),
-            storeOp.getMixedStrides(), storeOp.getDroppedDims(), resultOffsets,
-            resultSizes, resultStrides, combinedOffsets, combinedSizes,
-            combinedStrides))) {
-      return rewriter.notifyMatchFailure(
-          op, "failed to create tiled flow.dispatch.tensor.store op");
+    tensor::ExtractSliceOp sliceOp = currSliceOps.back();
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPoint(sliceOp);
```
Comment (pzread) [inline compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp:611]: Since the tiled producer will be used by multiple consumers, is it guaranteed that inserting at the last slice op will dominate all the uses?

**Comment by MaheshRavishankar** [comment]
@pzread based on your comments I changed the way the operations are fused. It now proceeds by pulling in producers "in-order" and then replacing all uses within already fused producers. This is related to how we decided we could do this in an offline conversation. PTAL.

**Comment by MaheshRavishankar** [comment]
Depends on https://reviews.llvm.org/D133039

File: compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp [hunk 5]
```diff
@@ -390,45 +478,193 @@ FailureOr<TilingResult> TileDispatchUsingSCFForOp::returningMatchAndRewrite(
...
+  // Get all ops to tile and fuse.
+  std::vector<Operation *> fusableProducers = getAllFusableProducers(op);
+  auto fusableProducersRef = llvm::makeArrayRef(fusableProducers);
+  while (!fusableProducersRef.empty()) {
+    auto fusableProducer = cast<TilingInterface>(fusableProducersRef.back());
+    fusableProducersRef = fusableProducersRef.drop_back();
+
+    /// Find a slice that is used to access the producer. Get all the slice ops.
+    /// It is assumed that the slice ops are returned in-order, so that you
+    /// could use the first slice as the insertion point.
```
Comment (pzread) [inline compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp:601]: Should be the last slice?

File: compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp [hunk 6]
```diff
@@ -390,45 +478,193 @@ FailureOr<TilingResult> TileDispatchUsingSCFForOp::returningMatchAndRewrite(
...
+    auto sliceOps = getAllFusableProducerUses(
+        fusableProducer, tileAndFuseResult.tiledAndFusedOps);
+    if (sliceOps.empty()) continue;
+
+    // Since slices are inserted in `candidateSlices` in BFS order, just use the
```
Comment (pzread) [inline compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp:606]: This comment looks irrelevant now (no longer have BFS and candidateSlices)

File: compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp [hunk 7]
```diff
@@ -390,45 +478,193 @@ FailureOr<TilingResult> TileDispatchUsingSCFForOp::returningMatchAndRewrite(
...
-                                        resultSizes))) {
-      return rewriter.notifyMatchFailure(
-          op, "failed to rewrite destructive update");
+
+  std::vector<Operation *> sortedOps = sortOpsTopologically(producers);
+  return sortedOps;
+}
+
+/// Return all slices that are used to access a tile of the producer. Assume
+/// that `tiledOps` are in "reverse" order of their appearance in the IR.
```
Comment (pzread) [inline compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp:560]: Maybe add a comment to indicate that the last slice op in the returned list will dominate all the other slice ops?

**Review by pzread — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp [hunk 8]
```diff
@@ -390,45 +478,193 @@ FailureOr<TilingResult> TileDispatchUsingSCFForOp::returningMatchAndRewrite(
...
+  }
+  // If there is no tiling then there is nothing to do for fusion.
+  if (!tilingResult->tiledLoops.any()) {
+    return tileAndFuseResult;
+  }
+  tileAndFuseResult.tiledAndFusedOps.push_back(tilingResult->tiledOp);
+  tileAndFuseResult.loops.append(tilingResult->loops);
+
+  // Get all ops to tile and fuse.
+  std::vector<Operation *> fusableProducers = getAllFusableProducers(op);
```
Comment (pzread) [inline compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp:593]: Since the `op` has been tiled by `tilingPattern.returningMatchAndRewrite(op, rewriter)`, I guess it has been erased? (So the ASAN crashed)
